### PR TITLE
Fix a bug affecting deeply nested function callbacks

### PR DESCRIPTION
**Pull request by joshessman-llnl**
# Summary

- This PR is a bugfix
- It does the following:
  - Fixes deeply nested function callbacks

This fix is conceptually similar to #449 - there's a lot of somewhat similar logic for iterating over nested objects and forwarding function calls to nested tables, so in the near future I'd like to look at refactoring that and keeping the iteration logic all in one place.

### CODE REVIEW

**Review by white238 — APPROVED** [review]

**Comment by white238** [comment]
LGTM

File: src/axom/inlet/tests/inlet_function.cpp [hunk 1]
```diff
@@ -442,6 +442,48 @@ TEST(inlet_function, dimension_dependent_result)
...
+  foo_schema.addFunction("bar",
+                         FunctionTag::Double,
+                         {FunctionTag::Double},
+                         "bar's description");
+
+  auto foos = inlet["quux"].get<std::vector<FooWithScalarFunc>>();
+  EXPECT_EQ(foos.size(), 2);
+
+  auto& first_func = foos[0].bar;
+  EXPECT_TRUE(static_cast<bool>(first_func));
```
Comment (kennyweiss) [inline src/axom/inlet/tests/inlet_function.cpp:479]: Is this a test that it's not a `nullptr`?  If so, that might more clearly show the intent.

**Review by kennyweiss — APPROVED** [review]

File: src/axom/inlet/tests/inlet_function.cpp (hunk 1 above)
Comment (joshessman-llnl) [reply src/axom/inlet/tests/inlet_function.cpp:479]: `std::function` doesn't make it easy to get the actual stored function pointer when it's wrapping a lambda, but I can add a comment clarifying what the check is doing.

**Comment by joshessman-llnl** [comment]
LGTM

**Pull request merged**